### ISSUE TITLE
Could shallow-backup integrate with git for backup?

### ISSUE THREAD

**Issue by jasikpark** [alichtman/shallow-backup#57]
It seems like it would be a great idea to make shallow-backup git aware and if you want it to back up to github.com or some other git remote repository when you run shallow-backup, then the backup is off-device, and it makes it easy to share.

Obviously, it's possible to do this manually, but it is also possible to reinstall all of the packages manually, and yet `$ pip install -r pip_list.txt` is already included in shallow-copy.

**Comment by alichtman** [comment]
Is it possible? Absolutely. I think it's a great idea, and something that I explored doing while I was building this. 

In terms of functionality, here's some pseudocode (in a mix of python and bash):

```shell
$ cd backup_dir

# Initialize git if not already initialized
if ".git/" does not exist in backup_dir:
     sp.call("git init", shell=True)

else:
    git add .
    git commit -m "Update shallow-backup"
    git push origin master
```

If you'd like to add this functionality, a PR would be welcomed! I could talk with you some more about how exactly to do this if you were interested.

> but it is also possible to reinstall all of the packages manually, and yet $ pip install -r pip_list.txt is already included in shallow-copy.

I'm confused what you meant by this. Can you rephrase it?

**Comment by jasikpark** [comment]
I was trying to make a comparison to make my case: between adding git
functionality and adding automatic package installation functionality. But
you already think it’s a great idea, so great!


On August 15, 2018 at 8:40:41 PM, Aaron Lichtman (notifications@github.com)
wrote:

Is it possible? Absolutely. I think it's a great idea, and something that I
explored doing while I was building this.

In terms of functionality, here's some pseudocode (in a mix of python and
bash):

$ cd backup_dir
# Initialize git if not already initializedif ".git/" does not exist
in backup_dir:
     sp.call("git init", shell=True)

else:
    git add .
    git commit -m "Update shallow-backup"
    git push origin master

If you'd like to add this functionality, a PR would be welcomed! I could
talk with you some more about how exactly to do this if you were interested.

but it is also possible to reinstall all of the packages manually, and yet
$ pip install -r pip_list.txt is already included in shallow-copy.

I'm confused what you meant by this. Can you rephrase it?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/alichtman/shallow-backup/issues/57#issuecomment-413395074>,
or mute the thread

**Comment by alichtman** [comment]
Got it, I understand. 

I'm totally behind that idea.

**Comment by pulkitmaloo** [comment]
I think this is a great idea as currently backing up to a git repository deletes all the files including the `.git` folder.

**Comment by alichtman** [comment]
Just opened a separate issue for that bug: #68

Once that is done, this shouldn't be too much work to complete.

**Comment by pulkitmaloo** [comment]
Thanks @alichtman. 

Also, is there an option to exclude private files? For example I noticed that it also backs up my private and public ssh-keys which I don't want to backup in the case where I'm pushing the files to a github repository. 

If I'm not wrong, there can be two solutions for this
1. Option to exclude files and don't back up those files at all
2. Add files to `.gitignore` so that they can be backed up on a local directory but will not be added to github repository.

I believe the first one can be implemented and second one user can do it by themselves.

**Comment by jasikpark** [comment]
It seems like it would be good to have reasonable defaults for preventing private information to be uploaded - the good thing is that Github already has software that removes private information from commits, but it would be better to never upload them.
